hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/HazelcastSqlSubstringFunction.java

```diff
@@ -0,0 +1,123 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.calcite.validate.operators;
+
+import com.google.common.collect.ImmutableList;
+import com.hazelcast.sql.impl.calcite.validate.types.ReplaceUnknownOperandTypeInference;
+import org.apache.calcite.linq4j.Ord;
+import org.apache.calcite.sql.SqlCall;
+import org.apache.calcite.sql.SqlCallBinding;
+import org.apache.calcite.sql.SqlFunction;
+import org.apache.calcite.sql.SqlFunctionCategory;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlOperandCountRange;
+import org.apache.calcite.sql.SqlUtil;
+import org.apache.calcite.sql.SqlWriter;
+import org.apache.calcite.sql.type.FamilyOperandTypeChecker;
+import org.apache.calcite.sql.type.OperandTypes;
+import org.apache.calcite.sql.type.ReturnTypes;
+import org.apache.calcite.sql.type.SqlOperandCountRanges;
+import org.apache.calcite.sql.type.SqlTypeFamily;
+import org.apache.calcite.sql.type.SqlTypeName;
+
+import java.util.List;
+
+import static org.apache.calcite.sql.type.SqlTypeName.INTEGER;
+import static org.apache.calcite.sql.type.SqlTypeName.VARCHAR;
+
+public class HazelcastSqlSubstringFunction extends SqlFunction {
+    public HazelcastSqlSubstringFunction() {
+        super(
+            "SUBSTRING",
+            SqlKind.OTHER_FUNCTION,
+            ReturnTypes.ARG0_NULLABLE_VARYING,
+            new ReplaceUnknownOperandTypeInference(new SqlTypeName[] { VARCHAR, INTEGER, INTEGER }),
+            null,
+            SqlFunctionCategory.STRING
+        );
+    }
+
+    @Override
+    public String getSignatureTemplate(int operandsCount) {
+        switch (operandsCount) {
+            case 2:
+                return "{0}({1} FROM {2})";
+            case 3:
+                return "{0}({1} FROM {2} FOR {3})";
+            default:
+                throw new AssertionError();
+        }
+    }
+
+    @Override
+    public String getAllowedSignatures(String opName) {
+        StringBuilder ret = new StringBuilder();
+        for (Ord<SqlTypeName> typeName : Ord.zip(SqlTypeName.CHAR_TYPES)) {
+            if (typeName.i > 0) {
+                ret.append(NL);
+            }
+
+            ret.append(
+                SqlUtil.getAliasedSignature(this, opName, ImmutableList.of(typeName.e, SqlTypeName.INTEGER))
+            );
+            ret.append(NL);
+            ret.append(
+                SqlUtil.getAliasedSignature(this, opName, ImmutableList.of(typeName.e, SqlTypeName.INTEGER, SqlTypeName.INTEGER))
+            );
+        }
+        return ret.toString();
+    }
+
+    @Override
+    public boolean checkOperandTypes(SqlCallBinding callBinding, boolean throwOnFailure) {
```

Comment:
notAny

Reply:
Fixed in https://github.com/hazelcast/hazelcast/pull/17379/commits/6a01e16bd9b3450ef82536e461db30bb9017f1bc